libbeat/processors/add_process_metadata/gosiger_cid_provider.go

```diff
@@ -0,0 +1,74 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package add_process_metadata
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type gosigarCidProvider struct {
+	hostPath           string
+	cgroupPrefixes     []string
+	processCgroupPaths func(string, int) (map[string]string, error)
+}
+
+func (p gosigarCidProvider) GetCid(pid int) (result string, err error) {
+	cgroups, err := p.processCgroupPaths(p.hostPath, pid)
```

Comment:
I've concerns about the performance here. In add_docker_metadata we have a cache only for this. That said, a cache has it's own problems in this context. I think we can go ahead without it and see how it goes cc @andrewkroh

Reply:
Same concerns from me. In Kubernetes, `pid` changes sometimes in a few seconds. I could add a cache the same as `add_docker_metadata`, but I personally would not use it.

Reply:
Without cache it would mean file read for each event, right. What are the downsides of simple cached pid-cid map like in my PR?

Reply:
Fetching this data each time can be expensive which is why we have a cache in add_docker_metadata. What are the problems with caching in this context? Is it with PID rollover/reuse? Short cache-times will mitigate that.

Reply:
The way I understand is that one parent process can spawn the child process, and the child process could be very short-lived. As @jtinkus mentioned before is the child may not exist by the time processor is checking the container id. And, I think this is the reason for the discussion about whether or not we check the `ppid` first or `pid`.

The concern I have is that when there are many pods running, and the processes on the pod spawn many short-lived children process. As there is PID rollover/reuse, it may enrich the event with the wrong container id.
![image](https://user-images.githubusercontent.com/24740574/76593689-3f78e580-64b4-11ea-9766-9564ed64e095.png)

Reply:
We can hit the disk io limitation before hitting this issue though after I think more about this.

Reply:
Maybe it is worth investigating a bit about the logic of PID reuse/rollover. There might be some mechanisms implemented to avoid getting same pids too often in short time for example. Just a guess of course...

Reply:
Let's not overthink this. So far this has worked well with `add_docker_metadata`. @BrookHF could you please add a similar cache to what we have in https://github.com/elastic/beats/blob/7eb2fba1f12e4bf0b623572d333ebc29e6f04202/libbeat/processors/add_docker_metadata/add_docker_metadata.go#L249-L266?

Reply:
other solution could be to talk to container runtime API and look for container ID.
https://stackoverflow.com/questions/24406743/coreos-get-docker-container-name-by-pid#24408480

E.g using docker sdk:
Fetch list of current containers:
https://godoc.org/github.com/docker/docker/client#Client.ContainerList
loop throught them to find the one match the pid
https://godoc.org/github.com/docker/docker/client#Client.ContainerInspect
or maybe this could be used but I pressume it'll work only for `docker exec`:
https://godoc.org/github.com/docker/docker/client#Client.ContainerExecInspect

We don't have to worry about cgroupsV2 nor cgroup prefixes since it'll be container runtime's responsibility to handle everything but I think it'd fall under `add_docker_metadata`.

Or we could use CRI api for it and aligned with all CRI compliant runtimes

Reply:
While that solution would work for Docker, we want other runtimes to be supported, so I don't think we can avoid cgroups for now. For the `add_docker_metadata` case, would that be solving a problem? as far as I know the current solution is working fine

Reply:
For `add_docker_metadata` it would solve a problem of `cgroup_prefixes` and detecting containers that were spawned by Docker on behave of Kubernetes.

I'm looking into CRI to see if it would solve our problem.

Reply:
Till there is no clear understanding how pid reuse works in containers/kubernetes, there is no way to say cgroup solution works fine. We might as well get certain % of wrongly assigned container ids :)

And still, starting with parent process, it should lower the % of not getting any container id and also lower % of wrongly assigned ones. In theory.

Reply:
cache times are configurable so I'm still in favor of a cache, you can always disable it by setting a really low timeout. We already have a cache for proc info here, we can just use the same timeouts, which makes sense.

@BrookHF I think this is the only pending bit, let me know if you need help with this, happy to send a commit to add it

Reply:
@exekiaa Thanks. I am actually working on the cache. Trying to make the expiration time and the cleanup time configurable also.